### ISSUE TITLE
better way to negate a variable referenced property

### ISSUE THREAD

**Issue by itaysk** [open-policy-agent/opa#2436]
If we want to assert that a variable referenced property doesn’t exists, we cannot intuitively negate it using `not`, instead we need to wrap it in a function/rule and negate that. For example:
This works:
```rego
there { input.list[_].prop }
```
This is expected to work but doesn't
```rego
notthere { not input.list[_].noprop }
```
This is the workaround
```rego
notthere { not checkthere } 
checkthere() { input.list[_].noprop }
```
[OPA Playground](https://play.openpolicyagent.org/p/WKtxBNfSwb)

I think this is a common use case, and the workaround is not great because it complicates otherwise concise and simple policies.
Can we consider supporting this use case in a language native, intuitive way? A basic solution could be to just apply the workaround by rewriting the rule for the user. I'm not familiar enough with rego's internals to suggest a better solution.

For reference: https://www.openpolicyagent.org/docs/latest/faq/#safety

**Comment by tsandall** [comment]
I agree that it would be nice if you could negate expressions containing variables that don't appear as outputs of other non-negated expressions however it raises semantic questions that do not have great answers. Importantly, it's unclear what the output of a query like `not input.list[_].noprop` should be. OPA can't generate all of the assignments to `_` because that set would be infinite.

I think there are really two cases here that we care about:

* Check if SOME element in list does not contain noprop
* Check if EVERY element in list does not contain noprop

The first can be expressed very easily today:

```
elem := input.list[_]
not elem.noprop
```

The second is harder and requires negation or comprehensions. This is one of the things that new users struggle with most in the language. We have an open issue to add a new keyword (`every`) to the language that would make this easier:

```
# SYNTAX TBD
every i {
  input.list[i]
}
not input.list[i].noprop
```

Basically you can define the _domain_ of the variable(s) (the bit between `{}`) and then express conditions that must be true for EVERY value in the domain.

Given that we have #2090 open to track this, I'm tempted to close unless you think this requires more discussion.

**Comment by itaysk** [comment]
Thanks. I would appreciate if you can confirm of correct my understanding of your answer: 
I was going to ask why you say the assignment set for `_` is infinite, but I think that's because I was thinking in terms of iterating a collection - that is a finite operation, and OPA is not working this way. So actually there was no real iteration in OPA, the same outcome is perceived when OPA tries to assign values to the indexing variable and happens to find all keys in the collection? And the "every" keyword is supposed to bring real iteration to the language?

**Comment by tsandall** [comment]
The reason there would be infinitely many assignments comes down to the language semantics.

The expression `input[i]` means `for some i, input[i] is defined and not false`. If input was `[true, false, 7]` then it would be defined for i=0, i=1, i=2 and not false for i=0 and i=2. The query `input[i]` with input as `[true, false, 7]` would produce two outputs: `[{i:0}, {i:2}]`.

OTOH, The expression `not input[i]` says `for some i, input[i] is not defined or false`. In this case, i=1 is is false so the query would produce that as an output. However, what about assignments to `i` that make `input[i]` undefined? Well, there are an infinite number of assignments that do that, e.g., i=3, i=4, i=-1, i="foo", etc.

Since we always want policy evaluation/analysis to terminate, queries that produce an infinite number of results are not allowed. We call queries that would produce an infinite number of results "unsafe". The safety checks in the compiler detect unsafe queries and reject them.

What `every` brings is the ability to change how variables are quantified. By default, variables are "existentially quantified"--queries that contain variables say "FOR SOME x, ...". With the `every` keyword, variables can be universally quantified--queries that contain universally quatified variables can say "FOR ALL x, ...".

Hope this helps!
